src/tree/context-dep.cc

```diff
@@ -178,9 +178,107 @@ void ContextDependency::Read (std::istream &is, bool binary) {
   to_pdf_ = to_pdf;
 }
 
-void ContextDependency::GetPdfInfo(const std::vector<int32> &phones,
-                                   const std::vector<int32> &num_pdf_classes,  // indexed by phone,
-                                   std::vector<std::vector<std::pair<int32, int32> > > *pdf_info) const {
+void ContextDependency::EnumeratePairs(
+    const std::vector<int32> &phones,
+    int32 self_loop_pdf_class, int32 forward_pdf_class,
+    const std::vector<int32> &phone_window,
+    unordered_set<std::pair<int32, int32>, PairHasher<int32> > *pairs) const {
+  std::vector<int32> new_phone_window(phone_window);
+  EventType vec;
+
+  std::vector<EventAnswerType> forward_pdfs, self_loop_pdfs;
+
+  int32 forward_pdf, self_loop_pdf;
+  if (Compute(phone_window, forward_pdf_class, &forward_pdf) &&
```

Comment:
Oh I see what you meant, with this code here, that's what prevented it from crashing, and it could be right.
But still, I think it would be better if you did it as I said-- that would be clearer, and may prevent unnecessary recursion.